File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    queue<int> bridge;
    int left_w = weight;
    
    for(int i = 0; i < bridge_length; i++)
        bridge.push(0);
    
    while(!truck_weights.empty()){
        answer++;
        auto it = truck_weights.begin();
        if(*it <= left_w) {
            bridge.pop();
            bridge.push(*it);
            left_w -= bridge.back();
            truck_weights.erase(it);
        }
        else {
            bridge.pop();
            bridge.push(0);
        }
        left_w += bridge.front();
    }
    
    return answer += bridge_length; //마지막 트럭이 빠지고 건너는 시간을 다리 길이만큼 더함.
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue<Integer> q = new LinkedList<>();
        
        for(int i = 0; i < bridge_length; i++)
            q.add(0);
        
        int n = truck_weights.length;
        int total = 0;
        int i = 0;
        
        if(bridge_length == 1)
            return n + 1;
        if(n == 1)
            return bridge_length + 1;
        
        while(i < n){
            total -= q.poll();
            answer++;
            
            if(total + truck_weights[i] <= weight){
                q.add(truck_weights[i]);
                total += truck_weights[i++];
            }else{
                q.add(0);
            }
        }
        
        
        return answer + bridge_length;
    }
}